adam_core/utils/indexable.py

```diff
@@ -1,174 +1,451 @@
+        self._class_index_to_members_is_slice : bool
+            Whether the mapping from the externally facing index to the class members are slices.
+        self._class_index_attribute : str, None
+            The attribute on which the index was set, if any. None if no attribute was used.
+        self._member_index : `~numpy.ndarray`
+            The index of the members of the class.
+        self._member_length : int
+            The length of the members of the class.
+
+
+        Raises
+        ------
+        ValueError: If all sliceable members do not have the same length.
+                    If the values are not None, str, or np.ndarray.
+        """
+        # Check if all the members are valid: have the same length
+        self._check_member_validity()
+
+        # Figure out the values that are going to be mapped to an index
+        # If no index values are given then we set the class index
+        # to be the range of the member lengths
+        if index_values is None:
+            class_index_values = np.arange(0, self._member_length)
+            self._class_index_attribute = None
 
-        return copy
+        else:
+            # Set the values that are to be indexed as a class index
+            if isinstance(index_values, str):
+                class_index_values = getattr(self, index_values)
+                self._class_index_attribute = index_values
 
-    def __delitem__(self, i: Union[int, slice, tuple, list, np.ndarray, pd.MultiIndex]):
-        ind = self._handle_index(i)
+            elif isinstance(index_values, np.ndarray):
+                class_index_values = index_values
+                self._class_index_attribute = None
 
-        for k, v in self.__dict__.items():
-            if isinstance(v, np.ma.masked_array):
-                self.__dict__[k] = np.delete(v, np.s_[ind], axis=0)
-                self.__dict__[k].mask = np.delete(v.mask, np.s_[ind], axis=0)
-            elif isinstance(v, np.ndarray):
-                self.__dict__[k] = np.delete(v, np.s_[ind], axis=0)
-            elif isinstance(v, Time):
-                self.__dict__[k] = Time(
-                    np.delete(v.mjd, np.s_[ind], axis=0), scale=v.scale, format="mjd"
-                )
-            elif isinstance(v, (list, Indexable)):
-                del v[ind]
-            elif isinstance(v, pd.MultiIndex):
-                self.__dict__[k] = v.delete(ind)
-            elif isinstance(v, UNSLICEABLE_DATA_STRUCTURES):
-                self.__dict__[k] = v
-            elif v is None:
-                pass
             else:
-                err = f"{type(v)} are not supported."
-                raise NotImplementedError(err)
+                err = "values must be None, str, or numpy.ndarray"
+                raise ValueError(err)
+
+        # If the index is to be set using an array that has an object or float dtype
+        # then we map the unique values of the index to integers. This will make querying the
```

Comment:
More accurately, if the array has a non-integer dtype.

Reply:
Fixed.